### ISSUE TITLE
Kernel/Fasta assumes default values are implicitly in a constant scope.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#32430]
Dart 2 constants generally don't need to write `const` in front if they are in a position where they are required to be constant: initializer of const variable, sub-expression of const expression, switch case expressions.
However, one location was deliberately omitted from that list: optional parameter default values. Those should not assume that they must be const, even though they currently must. We left this open so that we can later change default value behavior and allow non-constant values without accidentally making existing code behave differently. So, you still need to write `const` in front of a constant expression in a default-value position.

Example:
```dart
class C {
  final x;
  const C(this.x);
}
foo([y = C([])]) => y;
main() {
  print(foo().x);
}
```

This code should be a compile-time error because the default value of the `y` parameter must be constant, and it isn't.

This currently causes the following tests to fail: 
> implicit_creation/implicit_const_not_default_values_test/e{1..30}

**Comment by peter-ahe-google** [comment]
I think this can be fixed by converting `BodyBuilder.constantExpressionRequired` from a bool to an enum.

**Comment by peter-ahe-google** [comment]
Fixed in 59d859645ef0557bc7f8e2ee07957f506b755255.
